packages/flutter_tools/lib/src/web/devfs_web.dart

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+
+import 'package:meta/meta.dart';
+import 'package:mime/mime.dart' as mime;
+
+import '../base/common.dart';
+import '../base/file_system.dart';
+import '../base/io.dart';
+import '../build_info.dart';
+import '../convert.dart';
+import '../globals.dart';
+
+/// A web server which handles serving JavaScript and assets.
+///
+/// This is only used in development mode.
+class WebAssetServer {
+  @visibleForTesting
+  WebAssetServer(this._httpServer) {
+    _httpServer.listen((HttpRequest request) {
+      try {
+        return _handleRequest(request);
+      } catch (err) {
+        request.response.statusCode = HttpStatus.internalServerError;
+        return request.response.close();
+      }
+    });
```

Comment:
It would also be good to supply an onError callback here. It might do something like try to re-establish the connection. I'm not actually sure that would make sense, but an async unhandled exception sent to crash logging probably wouldn't be very informative.

Reply:
Added an onError callback that throwsToolExit by default. Tested that this can catch errors

Reply:
As discussed offline, `listen()` needs an `onError` callback to prevent errors on the listening socket (at least) from being propagated to the zone.

Reply:
Added an onError callback with a link to the dart SDK issue that prevents testing it